JSRT staticlib shutdown partial fix

jianchun opened the pull request:
A lot of shutdown/finalizer code expects current context. The current
staticlib shutdown atexit runs into problems. Since atexit is at process
shutdown, releasing memory isn't strictly required. This change reduces
current atexit to only shutdown threads. We'll investigate more in future
how to shutdown cleanly.

obastemur commented on lib/Jsrt/JsrtHelper.cpp:
```diff
@@ -89,40 +91,32 @@ void JsrtCallbackState::ObjectBeforeCallectCallbackWrapper(JsObjectBeforeCollect
 
 #ifndef _WIN32
-    static pthread_key_t s_threadLocalDummy;
+    static THREAD_LOCAL struct _OnThreadExit
```
This fails. Fix->separate the definition/declaration.

obastemur commented on lib/Jsrt/JsrtRuntime.cpp:
```diff
@@ -53,16 +53,20 @@ void JsrtRuntime::Uninitialize()
         currentThreadContext = currentThreadContext->Next();
 
+#ifdef CHAKRA_STATIC_LIBRARY
+        // TODO: Cleanup staticlib shutdown. This only shuts down threads.
```
nit: can we update `TODO` to `xplat-todo` ?

obastemur commented on lib/Runtime/Base/ThreadBoundThreadContextManager.cpp:
```diff
@@ -223,10 +223,19 @@ JsUtil::JobProcessor * ThreadBoundThreadContextManager::GetSharedJobProcessor()
+    bool deleteThreadContext = true;
+
+#ifdef CHAKRA_STATIC_LIBRARY
+    // TODO: Cleanup staticlib shutdown. Deleting contexts / finalizers having
```
nit: todo -> xplat-todo

obastemur commented on lib/Jsrt/JsrtHelper.cpp:
```diff
@@ -89,40 +91,32 @@ void JsrtCallbackState::ObjectBeforeCallectCallbackWrapper(JsObjectBeforeCollect
-            ThreadBoundThreadContextManager::DestroyContextAndEntryForCurrentThread();
-        }
+        atexit([]() {
+            ThreadBoundThreadContextManager::DestroyAllContextsAndEntries();
```
shall we check whether the current thread has an active context and that context isInScript?

jianchun replied:
It looks the added THREAD_LOCAL destructor is called before this which handled it already.

obastemur commented:
couple of nits and questions. 

LGTM

obastemur commented on lib/Jsrt/JsrtHelper.cpp:
```diff
@@ -89,40 +91,32 @@ void JsrtCallbackState::ObjectBeforeCallectCallbackWrapper(JsObjectBeforeCollect
 #endif
-    static THREAD_LOCAL bool s_threadWasEntered = false;
 
-    _NOINLINE void DISPOSE_CHAKRA_CORE_THREAD(void *_)
```
BTW; prev. pthread approach was guaranteed to run at the `last phase` of thread exit. Current approach doesn't have a priority in place. That makes me wonder that; If altering the timing for `ThreadBoundThreadContextManager::DestroyContextAndEntryForCurrentThread()` call has improved the test results positively, we might have a `race` issue on cleanup that we might have to target.

jianchun replied:
I don't know as it is hard to repro/investigate. (Some say pthread_key_create destructors might be skipped if pthread_exit is not called, e.g. by calling exit -- not sure if that plays a role).

jianchun commented:
@obastemur I reverted THREAD_LOCAL changes back to pthread_create_key, removed/tweaked a IsInScript() check and related code (we can do that because now I only shutdown threads), and now it looks works as well.

obastemur commented:
Great! :shipit:

jianchun commented:
Thanks @obastemur
